test_plans/SYCL-2020/aspect_selector.md

```diff
@@ -0,0 +1,127 @@
+
+### Function overloads
+
+Each test that tests aspects `{asp1, ..., aspN}` must test all three overloads of the `aspect_selector` function:
+
+* Pass a vector of aspects: `aspect_selector(std::vector{asp1, ..., aspN})`
+* Pass aspects as function arguments: `aspect_selector(asp1, ..., aspN)`
+* Pass aspects as template parameters: `aspect_selector<asp1, ..., aspN>()`
+
+### Combinations of aspects
+
+All aspects listed in the `enum class aspect` will be tested,
+let's identify these aspects as `{asp1, ..., aspN}`.
+
+The CTS will cover these combinations:
+
+1. All combinations where no aspect is passed in.
+2. All combinations where a single aspect is used.
+3. All combinations of two aspects,
+including duplicates (e.g. `aspect_selector(aspX, aspX)`),
+and reversed order (e.g. `aspect_selector(aspX, aspY)` and `aspect_selector(aspY, aspX)`).
+
+The CTS does not have to test all combinations of 3 or more aspects.
+
+The CTS will test all the different numbers of aspects at least once, up to `N`,
+where the aspects are not duplicated,
+i.e. `aspect_selector(asp1), aspect_selector(asp1, asp2), ..., aspect_selector(asp1, asp2, ..., aspN)`.
+
+Additional testing will include 100 combinations
+of randomly selected aspect combinations,
+allowing aspect duplicates,
+with a random arity.
+The random combinations are selected at test generation time,
+with each test generation producing the same combinations
+and each run testing the same random combinations.
+
+For each of the combinations described so far,
+add new combinations which also use forbidden aspects `{noasp1, ..., noaspM}` -
+`M` is independent of `N`, can be smaller, same size, or larger.
+These forbidden aspects will be passed to the `denyList` argument.
+Use the same rules as above for generating aspects,
+following these extra restrictions:
+
+* `noaspY` is not a member of the set `{asp1, ..., aspN}`
+    for each `noaspY` from the set of forbidden aspects.
+
+> Note: it is recommended a Python generator is written to cover all required combinations.
+
+## Tests
+
+### No aspects
+
+1. `auto selector1 = aspect_selector()`
+1. `auto selector2 = aspect_selector({})`
+1. `auto selector3 = aspect_selector({}, {})`
+1. `auto selector4 = aspect_selector<>()`
+
+In the test, all constructed selectors `selectorX` are used
+to construct a SYCL device, `devX` for each selector respectively.
+All devices `devX` must compare equal to the device obtained by `default_selector`
+(let's call this device `defaultDevice`).
+
+### Requested aspects using a device
+
+Assume requested aspects `{asp1, ..., aspN}` for each test,
+and forbidden aspects `denyList = {noasp1, ..., noaspM}`.
+
+These steps are performed before testing combinations:
+
+1. Iterate over all devices to find out if there's one that contains all requested aspects
+    but doesn't contain any forbidden ones,
+    we call this `suitableDevice`.
+
+    * If a device like this cannot be found, skip test.
```

Comment:
Why should the test be skipped in this case?  Instead, I think you could verify that the selector throws an `errc::runtime` exception as the spec mandates.